Avoid calling hasAttributeWithoutSynchronization() and then attributeWithoutSynchronization() for performance in HTMLAnchorElement.cpp

Ahmad-S792 opened the pull request:
#### a53f705a48ab788f88e35493ff5e878738387bdc
<pre>
Avoid calling hasAttributeWithoutSynchronization() and then attributeWithoutSynchronization() for performance in HTMLAnchorElement.cpp

<a href="https://bugs.webkit.org/show_bug.cgi?id=254751">https://bugs.webkit.org/show_bug.cgi?id=254751</a>

Reviewed by Chris Dumez and Aditya Keerthi.

Partial Merge - <a href="https://chromium.googlesource.com/chromium/blink/+/f28d0fba0e523c1669c48354a416a091c5f062e0">https://chromium.googlesource.com/chromium/blink/+/f28d0fba0e523c1669c48354a416a091c5f062e0</a>

It is to modify by avoid calling hasAttributeWithoutSynchronization and
attributeWithoutSynchronization for performance.
This causes the attribute to be looked up (linear search) twice instead
of once if we call &apos;attributeWithoutSynchronization&apos; directly and
then check if the returned is null.
Additionally, do early return if &apos;hyperlinkAuditingEnabled&apos; is not enabled.

* Source/WebCore/html/HTMLAnchorElement.cpp:
(HTMLAnchorElement::sendPings):

Canonical link: <a href="https://commits.webkit.org/262367@main">https://commits.webkit.org/262367@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a802924b3edaa0e857f5e3ae27ac29aebc273e47

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1333 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1372 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1416 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/2236 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/1188 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/1318 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1424 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/16/builds/1433 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/1342 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1345 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/1227 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/2082 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/1263 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/1212 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/1196 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/1218 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/1247 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/2318 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/1286 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/16/builds/1433 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/1200 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/1219 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/331 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1302 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/HTMLAnchorElement.cpp:
```diff
@@ -362,10 +362,11 @@ void HTMLAnchorElement::sendPings(const URL& destinationURL)
 
-    if (!hasAttributeWithoutSynchronization(pingAttr) || !document().settings().hyperlinkAuditingEnabled())
+    const auto& pingValue = attributeWithoutSynchronization(pingAttr);
+    if (pingValue.isNull() || !document().settings().hyperlinkAuditingEnabled())
```
Ideally, we'd do the hyperlinkAuditingEnabled check earlier to avoid even looking up the attribute when the setting is disabled.

Ahmad-S792 replied:
Should I add early return before const as below?
```
if (!document().settings().hyperlinkAuditingEnabled())
  return;
```

cdumez replied:
Yes please.

Ahmad-S792 replied:
Done. :-)

cdumez reviewed (approved):
LGTM

Ahmad-S792 committed:
Avoid calling hasAttributeWithoutSynchronization() and then attributeWithoutSynchronization() for performance in HTMLAnchorElement.cpp

https://bugs.webkit.org/show_bug.cgi?id=254751

Reviewed by Chris Dumez and Aditya Keerthi.

Partial Merge - https://chromium.googlesource.com/chromium/blink/+/f28d0fba0e523c1669c48354a416a091c5f062e0

It is to modify by avoid calling hasAttributeWithoutSynchronization and
attributeWithoutSynchronization for performance.
This causes the attribute to be looked up (linear search) twice instead
of once if we call 'attributeWithoutSynchronization' directly and
then check if the returned is null.
Additionally, do early return if 'hyperlinkAuditingEnabled' is not enabled.

* Source/WebCore/html/HTMLAnchorElement.cpp:
(HTMLAnchorElement::sendPings):

Canonical link: https://commits.webkit.org/262367@main
